internal/checks/promql_offset.go

```diff
@@ -0,0 +1,79 @@
+package checks
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"time"
+
+	promParser "github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/cloudflare/pint/internal/discovery"
+	"github.com/cloudflare/pint/internal/parser"
+)
+
+const (
+	OffsetCheckName = "promql/offset"
+)
+
+func NewOffsetCheck(prefix *TemplatedRegexp, minimumOffset time.Duration) OffsetCheck {
+	return OffsetCheck{prefixRegex: prefix, min: minimumOffset}
+}
+
+type OffsetCheck struct {
+	prefixRegex *TemplatedRegexp
+	min         time.Duration
+}
+
+func (c OffsetCheck) Meta() CheckMeta {
+	return CheckMeta{IsOnline: false}
+}
+
+func (c OffsetCheck) String() string {
+	return OffsetCheckName
+}
+
+func (c OffsetCheck) Reporter() string {
+	return OffsetCheckName
+}
+
+func (c OffsetCheck) Check(_ context.Context, _ string, rule parser.Rule, _ []discovery.Entry) (problems []Problem) {
+	expr := rule.Expr()
+
+	if expr.SyntaxError != nil || (rule.AlertingRule == nil && rule.RecordingRule == nil) {
+		return problems
+	}
+
+	for _, selector := range getSelectorsWithOffset(expr.Query) {
+		match, _ := regexp.MatchString(c.prefixRegex.anchored, selector.Name)
+		if match && c.min > selector.OriginalOffset {
+			problems = append(problems, Problem{
+				Fragment: selector.String(),
+				Lines:    expr.Lines(),
+				Reporter: c.Reporter(),
+				Text:     fmt.Sprintf("the %s metric requires a minimum offset of %s", selector.String(), c.min.String()),
+				Severity: Warning,
```

Comment:
I think we should allow the Severity to be dynamic.
Why did you go with Warning? For our use case, if it fails isn't it an error? Or are there edge cases?

Reply:
Maybe Kevin's worried about `aws_.*` metrics that are not from Cloudwatch. I don't know if there is any in our prod environment, worth checking. 

The alternative might be to check for specific label values (in our case, we have something like `config="yace"` or `config='cloudwatch-..."`), but that requires querying every single prom node again, so probably too much work/querying for little gain.

Reply:
This was my fear: that we'd be enforcing an offset with a regex match, but there is nothing in place to prohibit a dev from creating an `aws_.*` metric in the future that doesn't require an offset. 

I like the idea of querying for matching label values... but is that worth the extra querying time on Pint just to enforce an offset? I think the regex match will cover all cloud metrics, and in the above niche case we can override the validation.

Reply:
Agree the label checks aren't for now